cpp/src/arrow/util/macros.h

```diff
@@ -117,6 +125,7 @@
 #define ARROW_PREFETCH(addr)
 #define ARROW_RESTRICT
 #define ARROW_COMPILER_ASSUME(expr)
+#define ARROW_COMPILER_UNREACHABLE
```

Comment:
Should it use `https://en.cppreference.com/w/cpp/utility/unreachable` with an adequate feature test?